File: scripts/scraper-cycling-translate.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
scraper-cycling-translate.py
----------------------------
Traduit les descriptions + weather + publication_info du contenu scrapé cyclingstage.com
de EN vers FR (et autres langues) en utilisant le LLM z-ai.

Lit data/cycling/stage-favourites.json
Pour chaque stage, ajoute une clé "i18n" avec les traductions par langue.
Écrit data/cycling/stage-favourites-i18n.json

Usage:
    python3 scraper-cycling-translate.py                # traduit vers FR
    python3 scraper-cycling-translate.py --lang fr      # traduit vers FR
    python3 scraper-cycling-translate.py --lang fr,es   # traduit vers FR et ES
    python3 scraper-cycling-translate.py --force        # re-traduit même si déjà fait

Auteur: Équipe QA PARISCORE
"""

import argparse
import json
import os
import subprocess
import sys
import time

_SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
_REPO_DIR = os.path.dirname(_SCRIPT_DIR)
INPUT_FILE = os.path.join(_REPO_DIR, "data", "cycling", "stage-favourites.json")
OUTPUT_FILE = os.path.join(_REPO_DIR, "data", "cycling", "stage-favourites-i18n.json")

# Langues supportées par PARISCORE
SUPPORTED_LANGS = {
    'fr': 'French',
    'es': 'Spanish',
    'de': 'German',
    'it': 'Italian',
    'nl': 'Dutch',
    'pt': 'Portuguese',
}


def translate_text(text, target_lang_name, target_lang_code):
    """Traduit un texte EN vers la langue cible via le CLI z-ai.
    Retourne le texte traduit ou None si échec.
    """
    if not text or not text.strip():
        return None
    
    system_prompt = (
        f"You are a professional sports translator specializing in cycling. "
        f"Translate the following English text to {target_lang_name}. "
        f"Keep technical cycling terms accurate (e.g., 'team time trial', 'GC', 'yellow jersey', "
        f"'punchy climbers', 'gradient'). "
        f"Preserve the structure (paragraphs, line breaks). "
        f"Return ONLY the translated text, no commentary."
    )
    
    # Échappe les quotes pour la ligne de commande
    # On utilise un fichier temporaire pour éviter les problèmes d'échappement
    import tempfile
    with tempfile.NamedTemporaryFile(mode='w', suffix='.txt', delete=False, encoding='utf-8') as f:
        f.write(text)
        temp_path = f.name
    
    try:
        cmd = [
            'z-ai', 'chat',
            '-s', system_prompt,
            '-p', f"Translate this English cycling text to {target_lang_name}:\n\n{text}",
        ]
        
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=60
        )
        
        if result.returncode != 0:
            print(f"    [WARN] z-ai retour code {result.returncode}", file=sys.stderr)
            if result.stderr:
                print(f"    stderr: {result.stderr[:200]}", file=sys.stderr)
            return None
        
        # Parse la sortie (JSON ou texte)
        output = result.stdout.strip()
        
        # Le CLI z-ai ajoute des lignes "🚀 ..." avant le JSON
        # On extrait juste la partie JSON
        json_start = output.find('{')
        if json_start == -1:
            # Pas de JSON, retourne texte brut si valide
            if output and len(output) > 10 and not output.startswith('🚀'):
                return output
            return None
        
        # Trouve la fin du JSON (dernier })
        json_str = output[json_start:]
        # Enlève les éventuels caractères après le dernier }
        last_brace = json_str.rfind('}')
        if last_brace != -1:
            json_str = json_str[:last_brace + 1]
        
        try:
            data = json.loads(json_str)
            if isinstance(data, dict):
                # Format: {"choices": [{"message": {"content": "..."}}]}
                if 'choices' in data and data['choices']:
                    content = data['choices'][0].get('message', {}).get('content', '')
                    if content:
                        return content.strip()
        except json.JSONDecodeError:
            pass
        
        return None
        
    except subprocess.TimeoutExpired:
        print(f"    [WARN] Timeout traduction", file=sys.stderr)
        return None
    except Exception as e:
        print(f"    [WARN] Erreur traduction: {e}", file=sys.stderr)
        return None
    finally:
        try:
            os.unlink(temp_path)
        except:
            pass


def translate_stage(stage_data, lang_code, lang_name, force=False):
    """Traduit les champs pertinents d'un stage vers la langue cible.
    Ajoute une entrée stage_data['i18n'][lang_code] = {description, weather, ...}
    """
    if 'i18n' not in stage_data:
        stage_data['i18n'] = {}
    
    if not force and lang_code in stage_data['i18n']:
        print(f"    [SKIP] Déjà traduit en {lang_code}")
        return stage_data
    
    translations = {}
    
    # 1. Title
    if stage_data.get('title'):
        print(f"    [TRAD] title EN→{lang_code}...")
        translated = translate_text(stage_data['title'], lang_name, lang_code)
        if translated:
            translations['title'] = translated
    
    # 2. Description (le plus long)
    if stage_data.get('description'):
        print(f"    [TRAD] description EN→{lang_code}...")
        translated = translate_text(stage_data['description'], lang_name, lang_code)
        if translated:
            translations['description'] = translated
    
    # 3. Weather forecast
    if stage_data.get('weather_forecast'):
        print(f"    [TRAD] weather EN→{lang_code}...")
        translated = translate_text(stage_data['weather_forecast'], lang_name, lang_code)
        if translated:
            translations['weather_forecast'] = translated
    
    # 4. Publication info (court, pas critique)
    if stage_data.get('publication_info'):
        print(f"    [TRAD] publication_info EN→{lang_code}...")
        translated = translate_text(stage_data['publication_info'], lang_name, lang_code)
        if translated:
            translations['publication_info'] = translated
    
    if translations:
        stage_data['i18n'][lang_code] = translations
        print(f"    [OK] {len(translations)} champs traduits en {lang_code}")
    else:
        print(f"    [FAIL] Aucune traduction réussie")
    
    return stage_data


def main():
    parser = argparse.ArgumentParser(description="Traduit les favouris cyclingstage vers FR et autres langues")
    parser.add_argument('--lang', type=str, default='fr',
                        help='Code langue(s) cible(s), séparés par virgule (défaut: fr)')
    parser.add_argument('--force', action='store_true',
                        help='Re-traduit même si déjà présent')
    args = parser.parse_args()
    
    target_langs = [l.strip().lower() for l in args.lang.split(',') if l.strip()]
    
    # Valide les langues
    invalid = [l for l in target_langs if l not in SUPPORTED_LANGS]
    if invalid:
        print(f"ERREUR: langue(s) non supportée(s): {invalid}")
        print(f"Langues supportées: {list(SUPPORTED_LANGS.keys())}")
        sys.exit(1)
    
    print(f"=== TRADUCTION VERS {target_langs} ===")
    
    if not os.path.exists(INPUT_FILE):
        print(f"ERREUR: {INPUT_FILE} introuvable. Lance d'abord scraper-cyclingstage-favourites.py")
        sys.exit(1)
    
    # Charge les données
    with open(INPUT_FILE, 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    print(f"Stages à traduire: {len(data.get('stages', {}))}")
    print()
    
    # Pour chaque stage
    for stage_n, stage_data in data.get('stages', {}).items():
        if stage_data.get('status') != 'ok':
            print(f"[STAGE {stage_n}] SKIP (status={stage_data.get('status')})")
            continue
        
        print(f"[STAGE {stage_n}] {stage_data.get('title', '?')[:60]}...")
        
        for lang_code in target_langs:
            lang_name = SUPPORTED_LANGS[lang_code]
            translate_stage(stage_data, lang_code, lang_name, force=args.force)
            time.sleep(1)  # pause entre les langues
        
        print()
    
    # Sauvegarde
    os.makedirs(os.path.dirname(OUTPUT_FILE), exist_ok=True)
    data['i18n_last_update'] = time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())
    data['i18n_languages'] = target_langs
    
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)
    
    print(f"=== TERMINÉ ===")
    print(f"Fichier: {OUTPUT_FILE}")
    print(f"Stages traduits: {len(data.get('stages', {}))}")
    print(f"Langues: {target_langs}")


if __name__ == '__main__':
    main()

File: scripts/scraper-cycling-translate.js
#!/usr/bin/env node
// -*- coding: utf-8 -*-
/**
 * scraper-cycling-translate.js
 * ---------------------------
 * Port Node du scraper de traduction Python. Traduit le contenu EN scrapé
 * (data/cycling/stage-favourites.json) vers les langues supportées via Google Gemini,
 * et écrit data/cycling/stage-favourites-i18n.json (consommé par cyclingService.js).
 *
 * BUG ORIGINE (2026-07-12) : le script Python shell-out vers le CLI `z-ai` qui n'était
 * ni sur le PATH ni configuré (.z-ai-config absent) → le fichier i18n ne contenait
 * QUE 2 étapes avec des blocs i18n VIDES. Aucune traduction n'avait jamais réussi.
 * Porté en Node avec le SDK @google/generative-ai (déjà utilisé dans le projet pour
 * d'autres features IA) + retries/backoff + fusion incrémentale (on n'écrase jamais
 * une traduction existante sauf --force).
 *
 * Contrat de sortie (consommé par cyclingService.js:663 getStageFavourites) :
 *   stages["N"].i18n[lang] = { title?, description?, weather_forecast?, publication_info? }
 * Chaque champ est optionnel ; le service fallback champ par champ sur l'EN.
 *
 * Usage:
 *   node scraper-cycling-translate.js --lang fr                    # FR seulement, étapes manquantes
 *   node scraper-cycling-translate.js --lang fr,es,de,it,nl,pt     # 6 langues
 *   node scraper-cycling-translate.js --current --lang fr          # étape du jour seulement
 *   node scraper-cycling-translate.js --stage 9 --lang fr
 *   node scraper-cycling-translate.js --all --lang fr --force      # tout re-traduire
 *
 * Prérequis : GEMINI_API_KEY dans l'environnement ou le .env du projet.
 */
'use strict';

const fs = require('fs');
const path = require('path');

// ─── Constantes ───────────────────────────────────────────────────────────────
const SCRIPT_DIR = path.dirname(path.resolve(__filename));
const REPO_DIR = path.dirname(SCRIPT_DIR);
const INPUT_FILE = path.join(REPO_DIR, 'data', 'cycling', 'stage-favourites.json');
const OUTPUT_FILE = path.join(REPO_DIR, 'data', 'cycling', 'stage-favourites-i18n.json');
const CALENDAR_FILE = path.join(REPO_DIR, 'data', 'cycling', 'stages-calendar.json');

// gemini-2.0-flash-lite : quota free bien plus généreux que 2.5-flash pour les tâches
// textuelles simples comme la traduction. Override possible via GEMINI_MODEL.
const MODEL_NAME = process.env.GEMINI_MODEL || 'gemini-2.0-flash-lite';

// Langues supportées (cohérent avec l'objet I18N de pariscore.html:12359)
const SUPPORTED_LANGS = {
  fr: 'French',
  es: 'Spanish',
  de: 'German',
  it: 'Italian',
  nl: 'Dutch',
  pt: 'Portuguese',
};

const MAX_RETRIES = 5;
const RETRY_BASE_MS = 2000;       // backoff exponentiel : 2s, 4s, 8s, 16s, 32s
const RATE_LIMIT_WAIT_MS = 30000; // attente spécifique pour 429 Too Many Requests
const DELAY_BETWEEN_CALLS = 2000; // politesse entre appels Gemini (2s = ~30 req/min, sous le quota free)
const REQUEST_TIMEOUT_MS = 60000;

// ─── Chargement du .env (si pas déjà chargé par le processus parent) ──────────
// On lit manuellement .env pour ne pas dépendre du package dotenv (pas toujours installé).
function loadDotEnv() {
  const envPath = path.join(REPO_DIR, '.env');
  if (!fs.existsSync(envPath)) return;
  try {
    const lines = fs.readFileSync(envPath, 'utf8').split('\n');
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) continue;
      const eqIdx = trimmed.indexOf('=');
      if (eqIdx === -1) continue;
      const key = trimmed.slice(0, eqIdx).trim();
      let val = trimmed.slice(eqIdx + 1).trim();
      // Retire les quotes entourantes
      if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
        val = val.slice(1, -1);
      }
      if (key && !(key in process.env)) process.env[key] = val;
    }
  } catch (e) { /* ignore */ }
}
loadDotEnv();

// ─── Client Gemini (lazy init) ────────────────────────────────────────────────
let _genAI = null;
function getGenAI() {
  if (_genAI) return _genAI;
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    console.error('❌ ERREUR : GEMINI_API_KEY manquant.');
    console.error('   Ajoute GEMINI_API_KEY=... dans .env (ou exporte-le dans le shell).');
    console.error('   Obtiens une clé gratuite : https://aistudio.google.com/app/apikey');
    process.exit(2);
  }
  const { GoogleGenerativeAI } = require('@google/generative-ai');
  _genAI = new GoogleGenerativeAI(apiKey);
  return _genAI;
}

// ─── Prompt calibré cyclisme (repris du Python scraper-cycling-translate.py:52) ─
function buildSystemPrompt(langName) {
  return (
    'You are a professional sports translator specializing in cycling. ' +
    'Translate the following English text to ' + langName + '. ' +
    "Keep technical cycling terms accurate (e.g., 'team time trial', 'GC', 'yellow jersey', " +
    "'punchy climbers', 'gradient'). " +
    'Preserve the structure (paragraphs, line breaks). ' +
    'Return ONLY the translated text, no commentary, no preamble.'
  );
}

// ─── Traduction d'un texte via Gemini, avec retries ───────────────────────────
async function translateText(text, langCode) {
  if (!text || !text.trim()) return null;
  const langName = SUPPORTED_LANGS[langCode];
  if (!langName) throw new Error('Langue non supportée: ' + langCode);

  const genAI = getGenAI();
  const model = genAI.getGenerativeModel({
    model: MODEL_NAME,
    systemInstruction: buildSystemPrompt(langName),
  });

  const userPrompt = 'Translate this English cycling text to ' + langName + ':\n\n' + text;

  for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
    try {
      // AbortController pour respecter le timeout (au cas où le SDK hang)
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);
      const result = await model.generateContent({
        contents: [{ role: 'user', parts: [{ text: userPrompt }] }],
      }, { fetchOptions: { signal: controller.signal } });
      clearTimeout(timer);

      const response = result.response;
      const out = (response.text() || '').trim();
      if (!out) throw new Error('Réponse vide de Gemini');
      return out;
    } catch (e) {
      const msg = e.message || String(e);
      if (attempt < MAX_RETRIES) {
        // 429 Too Many Requests → attente plus longue (quota Gemini free limité)
        const isRateLimit = /429|Too Many|RESOURCE_EXHAUSTED|rate limit/i.test(msg);
        const wait = isRateLimit ? RATE_LIMIT_WAIT_MS : RETRY_BASE_MS * Math.pow(2, attempt - 1);
        const reason = isRateLimit ? 'RATE-LIMIT' : 'RETRY';
        console.error('    [' + reason + ' ' + attempt + '/' + MAX_RETRIES + '] ' + msg.slice(0, 90) + ' — attente ' + (wait / 1000) + 's');
        await new Promise((r) => setTimeout(r, wait));
      } else {
        console.error('    [FAIL] Traduction échouée après ' + MAX_RETRIES + ' essais: ' + msg.slice(0, 150));
        return null;
      }
    }
  }
  return null;
}

// ─── Helpers calendrier (pour --current) ──────────────────────────────────────
function loadCalendar() {
  const raw = JSON.parse(fs.readFileSync(CALENDAR_FILE, 'utf-8'));
  const map = {};
  for (const s of raw.stages) map[s.stage] = s.date;
  return map;
}
function determineCurrentStage() {
  const cal = loadCalendar();
  const today = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'Europe/Paris', year: 'numeric', month: '2-digit', day: '2-digit',
  }).format(new Date());
  for (const [n, date] of Object.entries(cal)) if (date === today) return Number(n);
  return null;
}

// ─── Lecture source + i18n existant ───────────────────────────────────────────
function loadInput() {
  return JSON.parse(fs.readFileSync(INPUT_FILE, 'utf-8'));
}
function loadExistingI18n() {
  if (fs.existsSync(OUTPUT_FILE)) {
    try { return JSON.parse(fs.readFileSync(OUTPUT_FILE, 'utf-8')); }
    catch (e) { console.error('[WARN] i18n existant illisible, on repart de zéro'); }
  }
  return null;
}

// ─── Traduction d'une étape pour une langue ───────────────────────────────────
// Champs à traduire (contrat cyclingService.js:663-668) :
const FIELDS = ['title', 'description', 'weather_forecast', 'publication_info'];

async function translateStage(stageData, langCode, force) {
  const existing = (stageData.i18n && stageData.i18n[langCode]) || {};
  const result = {};
  for (const field of FIELDS) {
    const sourceVal = stageData[field];
    if (!sourceVal || !String(sourceVal).trim()) continue; // pas de source → rien à traduire
    if (!force && existing[field]) {
      result[field] = existing[field]; // conserve la traduction existante
      continue;
    }
    const translated = await translateText(sourceVal, langCode);
    if (translated) {
      result[field] = translated;
      console.log('    [' + field + '] ✓ ' + translated.length + ' chars');
    } else {
      console.log('    [' + field + '] ✗ échec (conservé: ' + (existing[field] ? 'ancien' : 'rien') + ')');
      if (existing[field]) result[field] = existing[field]; // garde l'ancien si échec
    }
    await new Promise((r) => setTimeout(r, DELAY_BETWEEN_CALLS));
  }
  return result;
}

// ─── Écriture du fichier i18n (fusion, pas écrasement) ────────────────────────
function saveI18n(inputData, i18nData) {
  // On reconstruit le fichier i18n : recopie la source (toutes les étapes) + réinjecte les blocs i18n
  const out = JSON.parse(JSON.stringify(inputData)); // deep clone de la source (EN)
  for (const n of Object.keys(i18nData.stages || {})) {
    const src = out.stages[n];
    const tr = i18nData.stages[n];
    if (src && tr && tr.i18n) src.i18n = tr.i18n; // conserve le bloc i18n existant
  }
  out.i18n_last_update = new Date().toISOString();
  out.i18n_languages = Object.keys(SUPPORTED_LANGS);
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(out, null, 2), 'utf-8');
  console.log('[OK] i18n sauvegardé dans ' + OUTPUT_FILE);
}

// ─── CLI ──────────────────────────────────────────────────────────────────────
function findFirstMissingStage(inputData, existingI18n, langs) {
  // Retourne le numéro de la première étape (par ordre chronologique) dont la
  // traduction est incomplète pour au moins une des langues ciblées. Sert au mode
  // --next-missing utilisé par le cron : 1 étape par exécution pour rester sous le
  // quota Gemini free (au lieu d'une rafale de 500 requêtes qui déclenche des 429).
  const stages = Object.keys(inputData.stages || {}).map(Number).sort((a, b) => a - b);
  for (const n of stages) {
    const src = inputData.stages[String(n)];
    if (!src || src.status !== 'ok') continue;
    const i18nStage = (existingI18n.stages && existingI18n.stages[String(n)]) || {};
    const i18n = i18nStage.i18n || {};
    for (const lang of langs) {
      const tr = i18n[lang] || {};
      // Incomplet si un champ source non-vide n'a pas de traduction
      const incomplete = FIELDS.some((f) => src[f] && String(src[f]).trim() && !tr[f]);
      if (incomplete) return n;
    }
  }
  return null;
}

function parseArgs(argv) {
  const args = { langs: ['fr'], stage: null, all: false, current: false, nextMissing: false, force: false };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--all') args.all = true;
    else if (a === '--current') args.current = true;
    else if (a === '--next-missing') args.nextMissing = true;
    else if (a === '--force') args.force = true;
    else if (a === '--lang') { args.langs = (argv[++i] || '').split(',').map((s) => s.trim()).filter(Boolean); }
    else if (a.startsWith('--lang=')) args.langs = a.slice(7).split(',').map((s) => s.trim()).filter(Boolean);
    else if (a === '--stage') args.stage = Number(argv[++i]);
    else if (a.startsWith('--stage=')) args.stage = Number(a.slice(8));
    else if (a === '-h' || a === '--help') args.help = true;
  }
  // Validation des langues
  args.langs = args.langs.filter((l) => { if (!SUPPORTED_LANGS[l]) { console.error('Langue ignorée (non supportée): ' + l); return false; } return true; });
  if (!args.langs.length) { console.error('Aucune langue valide. Default: fr'); args.langs = ['fr']; }
  if (!args.stage && !args.all && !args.current && !args.nextMissing) args.all = true; // défaut: toutes les étapes
  return args;
}

async function main() {
  const args = parseArgs(process.argv);
  if (args.help) {
    console.log('Usage: node ' + path.basename(__filename) + ' [--lang fr,es,...] [--stage N | --current | --all] [--force]');
    console.log('Langues supportées: ' + Object.keys(SUPPORTED_LANGS).join(', '));
    process.exit(0);
  }

  console.log('=== Traduction Cyclisme TdF 2026 ===');
  console.log('Modèle Gemini : ' + MODEL_NAME);
  console.log('Langues       : ' + args.langs.join(', '));
  console.log('Force         : ' + (args.force ? 'oui (re-traduit tout)' : 'non (étapes/champs manquants seulement)'));
  console.log('');

  // Détermine les étapes à traiter
  const inputData = loadInput();
  const existingI18n = loadExistingI18n() || { stages: {} };
  if (!existingI18n.stages) existingI18n.stages = {};
  let stageNums;
  if (args.stage) stageNums = [args.stage];
  else if (args.current) {
    const n = determineCurrentStage();
    if (n === null) { console.log('[INFO] Pas d\'étape aujourd\'hui (jour de repos) — rien à faire.'); return; }
    stageNums = [n];
    console.log('[INFO] Étape du jour : ' + n);
  } else if (args.nextMissing) {
    // Mode cron : traduit UNE seule étape manquante (évite les 429 du quota free)
    const n = findFirstMissingStage(inputData, existingI18n, args.langs);
    if (n === null) { console.log('[INFO] Toutes les étapes sont déjà traduites pour les langues: ' + args.langs.join(',')); return; }
    stageNums = [n];
    console.log('[INFO] Mode next-missing : étape sélectionnée = ' + n);
  } else {
    stageNums = Object.keys(inputData.stages || {}).map(Number).sort((a, b) => a - b);
  }
  console.log('Étapes à traiter : ' + stageNums.join(', '));
  console.log('');

  // (existingI18n déjà chargé plus haut pour le mode --next-missing)

  // Init / sync les blocs stages du i18n avec la source (ajoute les étapes manquantes)
  for (const n of Object.keys(inputData.stages || {})) {
    if (!existingI18n.stages[n]) {
      existingI18n.stages[n] = JSON.parse(JSON.stringify(inputData.stages[n]));
      existingI18n.stages[n].i18n = existingI18n.stages[n].i18n || {};
    }
  }

  // Boucle principale
  let totalOk = 0;
  let totalFail = 0;
  for (const n of stageNums) {
    const stageData = inputData.stages[String(n)];
    if (!stageData) { console.log('[SKIP] Étape ' + n + ' absente de la source'); continue; }
    if (stageData.status && stageData.status !== 'ok') { console.log('[SKIP] Étape ' + n + ' status=' + stageData.status); continue; }
    const title = (stageData.title || '—').slice(0, 60);
    console.log('━━━ Stage ' + n + ': ' + title + ' ━━━');
    // Assure le bloc stage dans le i18n
    if (!existingI18n.stages[String(n)]) {
      existingI18n.stages[String(n)] = JSON.parse(JSON.stringify(stageData));
      existingI18n.stages[String(n)].i18n = {};
    }
    const i18nStage = existingI18n.stages[String(n)];
    i18nStage.i18n = i18nStage.i18n || {};

    for (const lang of args.langs) {
      console.log('  → ' + lang + ' (' + SUPPORTED_LANGS[lang] + '):');
      const translated = await translateStage(stageData, lang, args.force);
      if (translated && Object.keys(translated).length) {
        i18nStage.i18n[lang] = Object.assign(i18nStage.i18n[lang] || {}, translated);
        totalOk += Object.keys(translated).length;
      }
      // Sauvegarde incrémentale après chaque langue (pour ne pas perdre en cas de crash)
      saveI18n(inputData, existingI18n);
    }
  }

  console.log('');
  console.log('=== RÉSUMÉ ===');
  console.log('Champs traduits : ' + totalOk);
  console.log('Échecs          : ' + totalFail);
  console.log('Fichier         : ' + OUTPUT_FILE);
  console.log('i18n_last_update: ' + existingI18n.i18n_last_update);
}

main().catch((e) => { console.error('Erreur fatale:', e); process.exit(1); });
